File: server_integration_tests/servers/java_server_reactive/src/main/java/dev/graphlink/javaserver/impl/GreetingServiceImpl.java
package dev.graphlink.javaserver.impl;

import dev.graphlink.javaserver.generated.services.GreetingService;
import dev.graphlink.javaserver.generated.types.Greeting;
import graphql.GraphQLContext;

import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import reactor.core.publisher.Sinks;

@Service
public class GreetingServiceImpl implements GreetingService {

    private final Sinks.Many<Greeting> greetingReceivedSink = Sinks.many().multicast().directBestEffort();

    private Greeting buildGreeting(String name, String marker) {
        return Greeting.builder()
                .id("1")
                .message("Hello, " + name + (marker != null ? marker : ""))
                .build();
    }

    @Override
    public Mono<Greeting> greetIntercepted(String name, GraphQLContext context) {
        return Mono.just(buildGreeting(name, context.get("marker")));
    }

    @Override
    public Mono<Greeting> greetPlain(String name) {
        return Mono.just(buildGreeting(name, null));
    }

    @Override
    public Mono<Greeting> sendGreeting(String name) {
        Greeting greeting = buildGreeting(name, null);
        greetingReceivedSink.tryEmitNext(greeting);
        return Mono.just(greeting);
    }

    // `runBefore` (returning Mono<Void>, chained via .then()) has already
    // completed by the time this method runs, so the marker is captured
    // once, here, and baked into every subsequently emitted greeting for the
    // lifetime of this subscription.
    @Override
    public Flux<Greeting> greetingReceived(GraphQLContext context) {
        String marker = context.get("marker");
        return greetingReceivedSink.asFlux()
                .map(g -> marker != null
                        ? Greeting.builder().id(g.getId()).message(g.getMessage() + marker).build()
                        : g);
    }
}

File: server_integration_tests/servers/kotlin_server_suspend/src/main/kotlin/dev/graphlink/kotlinserversuspend/impl/GreetingServiceImpl.kt
package dev.graphlink.kotlinserversuspend.impl

import dev.graphlink.kotlinserversuspend.generated.services.GreetingService
import dev.graphlink.kotlinserversuspend.generated.types.Greeting
import graphql.GraphQLContext
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.map
import org.springframework.stereotype.Service

@Service
class GreetingServiceImpl : GreetingService {
    private val greetingReceivedFlow = MutableSharedFlow<Greeting>(extraBufferCapacity = 16)

    private fun buildGreeting(name: String, marker: String?): Greeting =
        Greeting(id = "1", message = "Hello, $name${marker ?: ""}")

    override suspend fun greetIntercepted(name: String, graphQLContext: GraphQLContext): Greeting =
        buildGreeting(name, graphQLContext.get("marker"))

    override suspend fun greetPlain(name: String): Greeting = buildGreeting(name, null)

    override suspend fun sendGreeting(name: String): Greeting {
        val greeting = buildGreeting(name, null)
        greetingReceivedFlow.tryEmit(greeting)
        return greeting
    }

    // `runBefore` has already completed by the time this method runs — the
    // generated controller calls it first — so the marker is captured once,
    // here, and baked into every subsequently emitted greeting for the
    // lifetime of this subscription.
    override fun greetingReceived(graphQLContext: GraphQLContext): Flow<Greeting> {
        val marker: String? = graphQLContext.get("marker")
        return greetingReceivedFlow.map { g -> if (marker != null) g.copy(message = g.message + marker) else g }
    }
}
